### PR TITLE
async version of uwsgi and http providers

### PR DESCRIPTION
**Pull request by oktopuz**
async version of uwsgi and http providers for the service autoscaler.

### CODE REVIEW

File: paasta_tools/autoscaling/autoscaling_service_lib.py [hunk 1]
```diff
@@ -182,19 +183,28 @@ def get_http_utilization_for_all_tasks(marathon_service_config, marathon_tasks,
...
             # This won't trigger in the event of DNS error or when a request is refused
             # a requests.exception.ConnectionError is raised in those cases
             utilization.append(1.0)
             log.debug("Recieved a timeout when querying %s on %s:%s. Assuming the service is at full utilization." % (
                 marathon_service_config.get_service(), task.host, task.ports[0]))
-        except Exception as e:
-            log.debug("Caught exception when querying %s on %s:%s : %s" % (
-                marathon_service_config.get_service(), task.host, task.ports[0], str(e)))
+        elif job.exception is not None:
+            log.debug("Caught exception when querying %s on %s:%s : %s" % (marathon_service_config.get_service(),
```
Comment (mattmb) [inline paasta_tools/autoscaling/autoscaling_service_lib.py]: I'd probably up this from debug, it is a real error even if it's not that critical.

File: paasta_tools/autoscaling/autoscaling_service_lib.py [hunk 2]
```diff
@@ -182,19 +183,28 @@ def get_http_utilization_for_all_tasks(marathon_service_config, marathon_tasks,
...
+            for task in marathon_tasks]
+    gevent.joinall(jobs)
+
+    for job in jobs:
+        if job.value is not None:
+            utilization.append(job.value)
+        elif isinstance(job.exception, requests.exceptions.Timeout):
             # If we time out querying an endpoint, assume the task is fully loaded
             # This won't trigger in the event of DNS error or when a request is refused
             # a requests.exception.ConnectionError is raised in those cases
```
Comment (mattmb) [inline paasta_tools/autoscaling/autoscaling_service_lib.py]: There's a subtle change here that means this comment isn't true now if I'm reading it right?

i.e. now if there's any exception querying a task we assume it to be "fully loaded". That might actually be reasonable although I think just dropping it from the list is also reasonable given that we are taking an average.

I guess I'm happy either way... but please update the comment if you want to keep the code like this.

Comment (oktopuz) [inline paasta_tools/autoscaling/autoscaling_service_lib.py]: Unless I missed something, I didn't change the way it worked.
A task should be considered as "fully loaded" when the job terminated with requests.exceptions.Timeout (tested with https://github.com/Yelp/paasta/blob/master/tests/autoscaling/test_autoscaling_service_lib.py#L312).
Any other exceptions should go to the line 203 (tested with https://github.com/Yelp/paasta/blob/master/tests/autoscaling/test_autoscaling_service_lib.py#L325).

Therefore the comment looks correct to me.

File: paasta_tools/autoscaling/autoscaling_service_lib.py (hunk 1 above)
Comment (oktopuz) [inline paasta_tools/autoscaling/autoscaling_service_lib.py]: replaced with log.error.

File: paasta_tools/autoscaling/autoscaling_service_lib.py (hunk 2 above)
Comment (mattmb) [inline paasta_tools/autoscaling/autoscaling_service_lib.py]: My bad, you're quite right! The behaviour is unchanged 👍

**Comment by mattmb** [comment]
Some test failures to deal with but otherwise this LGTM 😄

**Comment by oktopuz** [comment]
@mattmb, I had to redesign this a bit to keep the log messages in the same format.
Could you check it once again please?

**Comment by mattmb** [comment]
LGTM 👍

**Pull request merged**